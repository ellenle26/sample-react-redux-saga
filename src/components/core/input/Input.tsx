import classNames from "classnames";
import { FC } from "react";


const Input: FC<InputProps> = (props: InputProps) => {
    const required = props.required === true;
    const requiredMargin = required ? "mr-2" : "";
    const border = props.error ? "border-error" : "border-border";


    return (
        <div className={classNames("w-full", { "opacity-50": props.disabled }, `${props.className}`)}>
            {(props.title || required) && (
                <div className="mb-1 flex items-center text-sm font-bold">
                    {props.title && (
                        <p className={`${requiredMargin}`}>{props.title}</p>
                    )}
                </div>
            )}
            <input
                className={`w-full h-10 box-border border outline-none ${border} text-sm px-4 rounded-lg`}
                name={props.name}
                placeholder={props.placeholder}
                value={props.value}
                disabled={props.disabled ? props.disabled : false}
                onChange={(event) =>
                    props.onChange && props.onChange(event.target.value)
                }
            />
            {props.error && (
                <p className="text-xs mt-1 text-error">{props.error}</p>
            )}
        </div>
    );
};

interface InputProps {
    title?: string;
    name?: string;
    placeholder?: string;
    required?: boolean;
    className?: string;
    value?: any;
    error?: string;
    disabled?: boolean;
    onChange?: (value: any) => void;
}

export default Input;