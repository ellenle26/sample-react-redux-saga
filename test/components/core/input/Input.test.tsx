/**
 * @jest-environment jsdom
*/

import Input from "@components/core/input/Input";
import { fireEvent, render } from "@testing-library/react";
import React from "react";

test("snapshot testing", () => {
    const renderResult = render(<Input />);
    expect(renderResult.asFragment()).toMatchSnapshot();
});

test("change event testing", () => {
    const handleChange = jest.fn();
    const renderResult = render(<Input onChange={value => handleChange(value)} required title="title" error="error"/>);
    const inputTags = renderResult.container.getElementsByTagName("input");
    fireEvent.change(inputTags[0], { target: { value: "this is text" } });

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toBeCalledWith("this is text");
});

test('disabled testing', () => {
    const handleChange = jest.fn();
    const renderResult = render(<Input onChange={value => handleChange(value)} disabled error="error" />);
    const inputTags = renderResult.container.getElementsByTagName("input");
    expect(inputTags).toBeTruthy();
})