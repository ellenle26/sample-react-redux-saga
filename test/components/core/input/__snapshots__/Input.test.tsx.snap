// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="w-full undefined"
  >
    <input
      class="w-full h-10 box-border border outline-none border-border text-sm px-4 rounded-lg"
      value=""
    />
  </div>
</DocumentFragment>
`;
